control_plane/src/safekeeper.rs

```diff
@@ -116,81 +119,141 @@ impl SafekeeperNode {
+
+        let mut safekeeper_subprocess = filled_cmd.popen().with_context(|| {
+            format!("Safekeeper {id} failed to start. See console output and {safekeeper_log_path:?} for details.")
+        })?;
+        let safekeeper_pid = safekeeper_subprocess
+            .pid()
+            .with_context(|| format!("Safekeeper {id} subprocess got no pid"))?;
 
         // It takes a while for the safekeeper to start up. Wait until it is
         // open for business.
-        const RETRIES: i8 = 15;
-        for retries in 1..RETRIES {
-            match self.check_status() {
-                Ok(_) => {
+        for retries in 0..RETRIES {
+            match self.has_node_started(safekeeper_pid) {
+                Ok(true) => {
                     println!("\nSafekeeper started");
                     return Ok(());
                 }
-                Err(err) => {
-                    match err {
-                        SafekeeperHttpError::Transport(err) => {
-                            if err.is_connect() && retries < 5 {
-                                print!(".");
-                                io::stdout().flush().unwrap();
-                            } else {
-                                if retries == 5 {
-                                    println!() // put a line break after dots for second message
-                                }
-                                println!(
-                                    "Safekeeper not responding yet, err {} retrying ({})...",
-                                    err, retries
-                                );
-                            }
-                        }
-                        SafekeeperHttpError::Response(msg) => {
-                            bail!("safekeeper failed to start: {} ", msg)
+                Ok(false) => {
+                    if retries < 5 {
+                        print!(".");
+                        io::stdout().flush().unwrap();
+                    } else {
+                        if retries == 5 {
+                            println!() // put a line break after dots for second message
                         }
+                        println!("Safekeeper {id} has not started yet, retrying ({retries})...");
                     }
                     thread::sleep(Duration::from_secs(1));
                 }
+                Err(e) => {
+                    println!("Safekeeper {id} failed to start: {e:#}");
+                    if let Err(e) = safekeeper_subprocess.kill() {
+                        println!("Failed to stop safekeeper {id} subprocess: {e:#}")
+                    };
+                    return Err(e);
+                }
+            }
+        }
+        anyhow::bail!("safekeeper {id} failed to start in {RETRIES} seconds");
+    }
+
+    fn has_node_started(&self, expected_pid: u32) -> anyhow::Result<bool> {
+        match self.check_status() {
+            Ok(()) => {
+                let pid_file_path = self.pid_file();
```

Comment:
I don't think we need to read the pid file here. It's just a cross-check but it doesn't seem important to me, and less lines is generally better.